### ISSUE TITLE
Cannot use mean_absolute_error

### ISSUE THREAD

**Issue by jmlipman** [tensorflow/tensorflow#12252]
I made a very simple neural network on tf and I wanted to use mean absolute error loss function, however I got this error right after I created the optimizer:

No gradients provided for any variable, check your graph for ops that do not support gradients, between variables [...] and loss Tensor(...)

This is what I did:
```
cost = tf.metrics.mean_absolute_error(pred, y)[0]
optimizer = tf.train.AdadeltaOptimizer(learning_rate=learning_rate).minimize(cost)
```

I tried another loss function and it worked. In fact, I read that it's difficult to provide a gradient when the absolute value is involved, but I did exactly the same in Keras and it works. In fact, I also did the following (which is basically the mean absolute error) and it works as well!

```
cost = tf.reduce_mean(tf.abs(tf.subtract(pred, y)))
optimizer = tf.train.AdadeltaOptimizer(learning_rate=learning_rate).minimize(cost)
```

Why the function doesn't work?

JM.

**Comment by gissemari** [comment]
@jmlipman you might want to use another loss function like mean_squared_error for the optimization. Apart from that you can use mean_absolute_error to measure the "accuracy".

**Comment by truongnmt** [comment]
@gissemari can you please explain more detail how to implement `mean_absolute_error` to evaluate? 
CC @reedwm

**Comment by gissemari** [comment]
Sure. According to the example here, _cost_ can be used for the optimization, while _performance_ can be just the metric (as the module tf.metrics suggest).
cost = tf.losses.mean_squared_error(y, predictions)
optimizer = tf.train.AdadeltaOptimizer(learning_rate=learning_rate).minimize(cost)
performance  = tf.metrics.mean_absolute_error(pred, y)[0]

**Comment by EliasHasle** [comment]
Why do you say that mean absolute error has no gradient? It is piecewise smooth, just like a ReLU layer, and the gradient should be the mean of the gradients of each absolute error (well-defined everywhere except at 0). I was surprised not to find a mean_absolute_error in tf.losses, as right now I thought I could be better off training for that, because I suspect squared error is too strict for my data (perhaps encouraging the model to output the mean when it is uncertain, rather than risking a huge squared error by making a guess. Maybe I am wrong). Anyway, it should be possible to use `tf.reduce_mean(tf.abs(predictions-targets))`.
